nodejs/week2/src/backend/routes/meals.js

```diff
@@ -0,0 +1,123 @@
+  if (mealsIdOutput.length === 0) {
+    response.send("there are no meals with id: " + mealsIdString);
+  } else {
+    response.json(mealsIdOutput);
+  }
+});
+
+//The /meals route should support some query parameters:
+//maxPrice	Get meals that has a price smaller than maxPrice	Number Ex: /meals?maxPrice=90
+function getMealsSmallerThanMaxPrice(maxPrice) {
+  let mealsData = dataAccess.getDataFromJson(meals);
+  let maxPriceNumber = parseInt(maxPrice);
+  let mealsCheaperThanMaxPrice = mealsData.filter(
+    (meal) => meal.price <= maxPriceNumber
+  );
+
+  if (isNaN(maxPriceNumber)) {
+    result = "price should be written as a number";
+  } else if (mealsCheaperThanMaxPrice.length > 0) {
+    result = mealsCheaperThanMaxPrice;
+  } else {
+    result = `there are no meals less than ${maxPrice}`;
+  }
+  return result;
+}
+
+//title	Get meals that partially match a title. Rød grød med will match the meal with the title Rød grød med fløde	String
+// EX:	/meals?title=Indian%20platte
+//meals?title=Indian
+function getMealWithTheSameTitle(matchTitle) {
+  let mealsData = dataAccess.getDataFromJson(meals);
+  let mealsWithParticularTitle = mealsData.filter((meal) =>
+    meal.title.toLocaleLowerCase().includes(matchTitle.toLocaleLowerCase())
+  );
+  if (mealsWithParticularTitle.length === 0) {
+    return "there is no title with:" + matchTitle;
+  } else {
+    return mealsWithParticularTitle;
+  }
+}
+
+// created after date
+// Get meals that has been created after the date	Date	/meals?createdAfter=2019-04-05
+
+function getMaelsAfterDate(mealAfterDateString) {
+  let mealsData = dataAccess.getDataFromJson(meals);
+  let mealAfterDate = Date.parse(mealAfterDateString);
+  let mealsAfterDate = mealsData.filter(
+    ({ createdAt }) => Date.parse(createdAt) > mealAfterDate
+  );
+  if (mealsAfterDate.length === 0) {
+    return "there is no meal after: " + mealAfterDateString;
+  } else {
+    return mealsAfterDate;
+  }
+}
+
+// limit	Only specific number of meals	Date	/meals?limit=2
+
+function getSpecificCountOfMeals(mealLimitString) {
+  let mealsData = dataAccess.getDataFromJson(meals);
+  let mealLimit = parseInt(mealLimitString);
+  let mealsLimit = mealsData.slice(0, mealLimit);
+  return mealsLimit;
+}
+
+router.get("/meals", (request, response) => {
+  let result;
+  let mealLimitString = request.query.limit;
+  let mealAfterDateString = request.query.createdAfter;
+  let matchTitle = request.query.title;
+  let maxPrice = request.query.maxPrice;
+  let mealsData = dataAccess.getDataFromJson(meals);
+
+  if (mealLimitString) {
+    result = getSpecificCountOfMeals(mealLimitString);
+  } else if (mealAfterDateString) {
+    result = getMaelsAfterDate(mealAfterDateString);
+  } else if (matchTitle) {
+    result = getMealWithTheSameTitle(matchTitle);
+  } else if (maxPrice) {
+    result = getMealsSmallerThanMaxPrice(maxPrice);
+  } else if (mealsData) {
+    result = getAllMeals(meals);
+  }
```

Comment:
This only applies 1 query param. You need to be able to combine all of the query params to progressively filter the meals. 

In general, all of these query parameters are optional and could be provided or not with the request. If they are provided then you can use them to filter the meals according to maxPrice, title, etc. But if they are not provided you just don't need to use them. e.g.

/meals => give me all meals
/meals?maxPrice=100 => give me all cheap meals under 100 crowns
/meals?title=chicken => give me all meals with chicken
/meals?title=chicken&maxPrice=100 => give me all cheap meals with chicken

When the service behaves in this way in can be really useful because it provides a very rich functionality for the user.